xtarfile/formats/lz4.py

```diff
@@ -0,0 +1,71 @@
+from tarfile import ReadError, CompressionError
+
+
+# If you want to extend with additional formats use this file as an example!
+
+
+# This variable is used to register new compression functions in the Tarfile class
+# Dictionary format: 'file suffix' : 'open function'
+compdict = {'lz4': 'lz4open'}
+
+
+# This variable is used to overload _StreamProxy.getcomptype with additional formats
+# Make sure indentation is correct and that the line ends with '\n'
+getcomptype = ['    elif self.buf.startswith(b"\\x04\\x22\\x4D\\x18"): return "lz4"\n']
+
+
+# This variable is used to overload _Stream.__init__ with additional formats
+# Make sure indentation is correct and that all lines end with '\n'
+streaminit = ['        elif comptype == "lz4":\n',
+              '            try:\n',
+              '                from lz4.frame import LZ4FrameCompressor, LZ4FrameDecompressor\n',
+              '            except ImportError:\n',
+              '                raise CompressionError("lz4 module is not available")\n',
+              '            if mode == "r":\n',
+              '                self.dbuf = b""\n',
+              '                self.cmp = LZ4FrameDecompressor()\n',
+              '                self.exception = RuntimeError\n',
+              '            else:\n',
+              '                self.cmp = LZ4FrameCompressor()\n',
+              '                self.fileobj.write((self.cmp.begin()))\n']
```

Comment:
This approach is amazingly clever, kudos for that. At the same time it's also scary since it seems potentially hard to maintain and a little brittle, e.g. the `exit(1)` code below checking for specific line numbers of the implementation in the standard library.

As an in-between the fully auto-magic approach suggested here and the approach in the code at my last review where all the various formats were handled inline in a single function, would something like the following work?

1. Each file in the formats folder exposes a `stream_init(self, name, mode, comptype, fileobj, bufsize)` method which implements the required stream initialization code.
2. Each file in the formats folder exposes a `comptype` tuple of compression formats that it can accept.
3. In the base xtarfile module, either dynamically load all the format implementations, or have a tuple that explicitly enumerates the modules.
4. In the base xtarfile module, add a new function which iterates through the available format modules, checks which of them accepts the current `comptype` and then calls the `stream_init` function of the module.

Sample pseudo-code below to illustrate the idea above:

```py
# in xtarfile.py
_base_stream_init = _Stream.__init__

def _customized_stream_init(self, name, mode, comptype, fileobj, bufsize):
    for module in loaded_format_modules:
        if comptype in module.comptypes:
            module.stream_init(self, name, mode, comptype, fileobj, bufsize)
            break
    else:
        # if we get here, none of the custom formats can handle this so fallback to the built-in ones
        _base_stream_init(self, name, mode, comptype, fileobj, bufsize)

_Stream.__init__ = _customized_stream_init

# in formats/zstandard.py

comptypes = ('zst', 'zstd')

def stream_init(self, name, mode, comptype, fileobj, bufsize):
    try:    
        import zstandard    
    except ImportError:    
        raise CompressionError("lzma module is not available")    
    if mode == "r":    
        self.dbuf = b""    
        dctx = zstandard.ZstdDecompressor()    
        self.cmp = dctx.decompressobj()    
        self.exception = zstandard.ZstdError    
    else:    
        dctx = zstandard.ZstdCompressor()    
        self.cmp = dctx.compressobj()
```

The advantage of this approach is that we keep all the code in source Python as opposed to compiling it from strings. A similar approach may also work for `getcomptype`. Thoughts?

Reply:
@c-w Not using strings but actual functions for the inserted code is a good idea and should be easy to resolve.

I'm not sure about your other ideas though. Maybe I'm just not smart enough to understand fully.
A minimal `_Stream.__init__` for **zst** would look like this.
```python
def stream__init__(self, name, mode, comptype, fileobj, bufsize):
    """Construct a _Stream object.
    """
    self._extfileobj = True
    if fileobj is None:
        fileobj = _LowLevelFile(name, mode)
        self._extfileobj = False

    if comptype == '*':
        # Enable transparent compression detection for the
        # stream interface
        fileobj = _StreamProxy(fileobj)
        comptype = fileobj.getcomptype()

    self.name     = name or ""
    self.mode     = mode
    self.comptype = comptype
    self.fileobj  = fileobj
    self.bufsize  = bufsize
    self.buf      = b""
    self.pos      = 0
    self.closed   = False

    try:
        if comptype in ("zst", "zstd"):
            try:
                import zstandard
            except ImportError:
                raise CompressionError("lzma module is not available")
            if mode == "r":
                self.dbuf = b""
                self.cmp = zstandard.ZstdDecompressor().decompressobj()
                self.exception = zstandard.ZstdError
            else:
                self.cmp = zstandard.ZstdCompressor().compressobj()

        else:
            raise CompressionError("unknown compression type %r" % comptype)

    except:
        if not self._extfileobj:
            self.fileobj.close()
        self.closed = True
        raise
```
And suffers from the same issues as just inserting the code directly into `_Stream.__init__`. Meaning it will require a review every single time **tarfile.py** is changed.
I do not see anyway to get around needing a review for every single change to **tarfile.py**. At least if you want 100% equal behavior. Hell, the entire PR relies on the internal structure of **tarfile.py** so there is no getting around it :D
This entire PR would require significantly more maintenance than the previous solution. Even if you skip the streaming mode. Maybe it's just not a good idea at all.